ako-gateway-api/nodes/gateway_model.go

```diff
@@ -179,86 +176,56 @@ func BuildVsVipNodeForGateway(gateway *gatewayv1.Gateway, vsName string) *nodes.
-		if akogatewayapilib.IsListenerInvalid(gwStatus, i) {
-			invalidListeners += 1
+	encodedCertName := akogatewayapilib.GetTLSKeyCertNodeName(gateway.Namespace, gateway.Name, certNamespace, secretName)
+	for evhVsCertRef := range evhVsCertRefs {
+		if evhVsCertRefs[evhVsCertRef].Name != encodedCertName {
+			tlsNodes = append(tlsNodes, evhVsCertRefs[evhVsCertRef])
 		}
 	}
-	if invalidListeners < totalListeners && len(tlsNodes) > 0 {
+	if len(tlsNodes) > 0 {
 		object.GetAviEvhVS()[0].SSLKeyCertRefs = tlsNodes
 	} else {
 		utils.AviLog.Warnf("key: %s, msg: No certificate present for Parent VS %s", key, object.GetAviEvhVS()[0].Name)
 		object.GetAviEvhVS()[0].SSLKeyCertRefs = nil
 	}
 	utils.AviLog.Infof("key: %s, msg: Updated cert_refs in parentVS: %s", key, object.GetAviEvhVS()[0].Name)
-	return invalidListeners == totalListeners
+	return akogatewayapilib.IsGatewayInvalid(gwStatus)
 }
 
-func AddTLSNode(key string, object *AviObjectGraph, gateway *gatewayv1.Gateway, secretObj *corev1.Secret, encodedCertNameIndexMap map[string][]int) {
-	var tlsNodes []*nodes.AviTLSKeyCertNode
-	_, _, secretName := lib.ExtractTypeNameNamespace(key)
-	evhVsCertRefs := object.GetAviEvhVS()[0].SSLKeyCertRefs
+func AddTLSNode(key string, object *AviObjectGraph, gateway *gatewayv1.Gateway, secretObj *corev1.Secret) {
+	_, certNamespace, secretName := lib.ExtractTypeNameNamespace(key)
+	tlsNodes := object.GetAviEvhVS()[0].SSLKeyCertRefs
 	gwStatus := akogatewayapiobjects.GatewayApiLister().GetGatewayToGatewayStatusMapping(gateway.Namespace + "/" + gateway.Name)
+	foundMatchingCertRef := false
 	for i, listener := range gateway.Spec.Listeners {
 		if akogatewayapilib.IsListenerInvalid(gwStatus, i) {
 			continue
 		}
 		if listener.TLS != nil {
 			for _, certRef := range listener.TLS.CertificateRefs {
 				name := string(certRef.Name)
-				// Hostname can be empty so initialize it to section name of listener
-				hostname := string(listener.Name)
-				if listener.Hostname != nil {
-					hostname = string(*listener.Hostname)
+				listenerCertRefNamespace := gateway.Namespace
+				if certRef.Namespace != nil {
+					listenerCertRefNamespace = string(*certRef.Namespace)
 				}
-				encodedCertName := lib.GetTLSKeyCertNodeName("", hostname, name)
-				indexlist, exists := encodedCertNameIndexMap[encodedCertName]
-				if exists {
-					tlsNodes = append(tlsNodes, evhVsCertRefs[indexlist[0]])
-					if len(indexlist) > 1 {
-						encodedCertNameIndexMap[encodedCertName] = indexlist[1:]
-					} else {
-						delete(encodedCertNameIndexMap, encodedCertName)
-					}
-				} else {
-					if name == secretName {
-						tlsNode := TLSNodeFromSecret(secretObj, hostname, name, key)
-						tlsNodes = append(tlsNodes, tlsNode)
-					}
+				if name == secretName && listenerCertRefNamespace == certNamespace {
+					tlsNode := TLSNodeFromSecret(secretObj, gateway.Namespace, gateway.Name, certNamespace, secretName, key)
+					tlsNodes = append(tlsNodes, tlsNode)
```

Comment:
in case of secret update we might have 2 certs attached to parent vs. (both old and new certs)

Reply:
Done